### PR TITLE
Consolidate logger to model agent sidecar and add readiness probe to model container

### LINKED ISSUE #1124: Unexpected (false-positive) model ready status when using batcher
/kind bug

Usually when deploying kfserving endpoint, the inferenceservices won't be marked as ready=True (can be monitored through `kubectl get inferenceservices` or python sdk `KFServing.get(watch=True)`) until the model gets ready (i.e., warmup finished through load function). However, this is not the case when using batcher: the inferenceservices will be marked as ready as soon as the batcher sidecar is ready and all containers running, before the model warmup finished. So in case there's a bug in model load function, if using batcher, we'll get false-positive and thus can't get the correct ready status.

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

Here I created a simple custom model docker (model will never be ready) and uploaded to `lplrobert/kfserving-not-ready`

```
# Dockerfile
FROM python:3.6-alpine

WORKDIR /app

COPY app.py /app

RUN apk update
RUN apk add make automake gcc g++ subversion python3-dev libffi-dev openssl-dev

RUN pip install kfserving==0.4.0

CMD ["python", "app.py"]
```

```
# app.py
import kfserving
import time

class KFServingSampleModel(kfserving.KFModel):
    def __init__(self, name):
        super().__init__(name)
        self.ready = False

    def load(self):
        time.sleep(60)
        raise RuntimeError("Model load failed!")

    def predict(self, request):
        return request

if __name__ == "__main__":
    model = KFServingSampleModel("kfserving-custom-model")
    model.load()
    kfserving.KFServer(workers=1).start([model])
```

```
# custom.yaml
apiVersion: serving.kubeflow.org/v1alpha2
kind: InferenceService
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: kfserving-custom-model
spec:
  default:
    predictor:
      batcher:
        maxBatchSize: 32
        maxLatency: 5000
        timeout: 60
      custom:
        name: kfserving-custom-model
        container:
          image: lplrobert/kfserving-not-ready
```

If we apply above CRD (with batcher enabled) and watch at `kubectl get inferenceservices`, we'll see `kfserving-custom-model` gets ready pretty quickly
```
NAME                              URL                                                              READY   DEFAULT TRAFFIC   CANARY TRA
FFIC   AGE
kfserving-custom-model            http://kfserving-custom-model.namespace.example.com            True    100
       15s
```

Then become unready only after waiting for a few (>2) minutes when the pod restarted and failed
```
NAME                              URL                                                              READY   DEFAULT TRAFFIC   CANARY TRA
FFIC   AGE
kfserving-custom-model                                                                             False
       2m15s
```

Then become ready again quickly when the pod retry again
```
NAME                              URL                                                              READY   DEFAULT TRAFFIC   CANARY TRA
FFIC   AGE
kfserving-custom-model            http://kfserving-custom-model.namespace.example.com            True    100
       2m27s
```

And after that, the ready status will change back and forth when the pod fails and restarts. In that case, we'll be misled to think the model works fine at first and can't be alerted until waiting for quite a long time.

While in comparison, if we apply the above CRD with batcher section removed. The inferenceservice will never be in ready status, and thus we know at first there's a bug somewhere.

**What did you expect to happen:**
If using batcher, inferenceservice should also return ready=True only if model warmup succeed, rather than when all containers are running.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Istio Version:
- Knative Version:
- KFServing Version: 0.4.0
- Kubeflow version:
- Kfdef:[k8s_istio/istio_dex/gcp_basic_auth/gcp_iap/aws/aws_cognito/ibm]
- Minikube version:
- Kubernetes version: (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by yuzisun**
**What this PR does / why we need it**:
- Consolidate logger and model puller to single sidecar agent
- Add readiness probe for agent to ensure main model container health as queue proxy now probes the agent sidecar

**Which issue(s) this PR fixe** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1124 

**Special notes for your reviewer**:

1. Remove logger binary and consolidate into agent image

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by yuzisun** [comment]
@cliveseldon Can you review this, this adds the readiness probe for the logger and consolidates to a single sidecar.

File: cmd/agent/main.go [hunk 1]
```diff
@@ -1,29 +1,238 @@
...
+	healthState := &health.State{}
+	ctx := signals.NewContext()
+	// Setup probe to run for checking user-application healthiness.
+	probe := buildProbe(logger, env.ServingReadinessProbe, *componentPort)
+	mainServer := buildServer(ctx, *port, *componentPort, loggerArgs, healthState, probe, logger)
+	servers := map[string]*http.Server{
+		"main": mainServer,
+	}
+	errCh := make(chan error)
+	listenCh := make(chan struct{})
```
Comment (cliveseldon) [inline cmd/agent/main.go:120]: Should we reference the copy of code from https://github.com/knative/serving/blob/master/cmd/queue/main.go ?

Comment (yuzisun) [reply cmd/agent/main.go:120]: sure

**Pull request merged**